File: Novels/packages/6-mobile-android/StoryReader/app/src/main/java/com/storysphere/storyreader/tts/TextToSpeechManager.kt
package com.storysphere.storyreader.tts

import android.content.Context
import android.util.Log
import javax.inject.Inject
import javax.inject.Singleton

enum class TTSEngineType {
    NATIVE,
    EMBEDDED // Proprietary SDK (60MB, high quality)
}

@Singleton
class TextToSpeechManager @Inject constructor(
    private val context: Context,
    private val nativeTTSEngine: NativeTTSEngine
) {
    private var currentEngine: TTSEngine? = null
    private var engineType: TTSEngineType = TTSEngineType.NATIVE
    
    init {
        // Default to native TTS
        currentEngine = nativeTTSEngine
    }
    
    fun setEngine(type: TTSEngineType) {
        engineType = type
        currentEngine = when (type) {
            TTSEngineType.NATIVE -> nativeTTSEngine
            TTSEngineType.EMBEDDED -> {
                // TODO: Initialize embedded TTS engine when available
                Log.w("TextToSpeechManager", "Embedded TTS engine not yet implemented")
                nativeTTSEngine
            }
        }
    }
    
    suspend fun speak(text: String, onUtteranceCompleted: (() -> Unit)? = null) {
        currentEngine?.speak(text, onUtteranceCompleted)
    }
    
    suspend fun stop() {
        currentEngine?.stop()
    }
    
    suspend fun pause() {
        currentEngine?.pause()
    }
    
    suspend fun resume() {
        currentEngine?.resume()
    }
    
    fun setSpeechRate(rate: Float) {
        currentEngine?.setSpeechRate(rate)
    }
    
    fun setPitch(pitch: Float) {
        currentEngine?.setPitch(pitch)
    }
    
    fun isSpeaking(): Boolean {
        return currentEngine?.isSpeaking() ?: false
    }
    
    fun shutdown() {
        currentEngine?.shutdown()
        currentEngine = null
    }
}

File: Novels/packages/5-mobile-ios/StoryReader/Service/TTS/TextToSpeechManager.swift
import Foundation
import Combine
import AVFoundation

// TTS Manager - Delegates to appropriate engine based on user settings
class TextToSpeechManager: ObservableObject {
    static let shared = TextToSpeechManager()
    
    @Published var isSpeaking: Bool = false
    @Published var currentProgress: Double = 0.0
    @Published var selectedEngine: TTSEngineType = .native
    @Published var selectedVoice: String?
    @Published var speed: Double = 1.0
    
    private var currentEngine: TTSEngine {
        switch selectedEngine {
        case .native:
            return nativeEngine
        case .embedded:
            return embeddedEngine
        }
    }
    
    private let nativeEngine = NativeTTSEngine()
    private let embeddedEngine = EmbeddedTTSEngine()
    
    init() {
        // Engines are initialized
    }
    
    // Synthesize speech (for audio file generation)
    func synthesize(text: String) async throws -> Data {
        return try await currentEngine.synthesize(text: text, voice: selectedVoice, speed: speed)
    }
    
    // Speak text (for immediate playback)
    func speak(text: String) async throws {
        try await currentEngine.speak(text: text, voice: selectedVoice, speed: speed)
        isSpeaking = currentEngine.isSpeaking
        currentProgress = currentEngine.currentProgress
    }
    
    func stop() {
        currentEngine.stop()
        isSpeaking = false
        currentProgress = 0.0
    }
    
    func pause() {
        currentEngine.pause()
        isSpeaking = false
    }
    
    func resume() {
        currentEngine.resume()
        isSpeaking = true
    }
    
    // Get available voices
    func getAvailableVoices() -> [AVSpeechSynthesisVoice] {
        return AVSpeechSynthesisVoice.speechVoices()
    }
}

enum TTSEngineType: String, Codable {
    case native = "native" // AVSpeechSynthesizer
    case embedded = "embedded" // Proprietary SDK (60MB)
}

